R/resultAnnotations.R

```diff
@@ -0,0 +1,925 @@
+#'   fds <- createTestFraserDataSet()
+#'   
+#'   # load reference annotation
+#'   requireNamespace("TxDb.Hsapiens.UCSC.hg19.knownGene")
+#'   txdb <- 
+#'     TxDb.Hsapiens.UCSC.hg19.knownGene::TxDb.Hsapiens.UCSC.hg19.knownGene 
+#'     
+#'   # add basic annotations for overlap with the reference annotation
+#'   # run this function before creating the results table
+#'   fds <- annotateIntronReferenceOverlap(fds, txdb) 
+#' 
+#'   # extract results: for this small example dataset, only a z score cutoff 
+#'   # of 1 is used to get at least one result. 
+#'   # Make sure to include the additional column in the results table
+#'   res <- results(fds, padjCutoff=NA, zScoreCutoff=1, deltaPsiCutoff=NA, 
+#'                   additionalColumns = 'annotatedJunction')
+#'   res_dt <- as.data.table(res)
+#'   
+#'   # annotate the type of splice event and UTR overlap
+#'   res_dt <- annotateSpliceEventType(result=res_dt, txdb=txdb, fds=fds)
+#'   
+#'   # annotate overlap with blacklist regions
+#'   res_dt <- flagBlacklistRegions(result=res_dt, assemblyVersion="hg19")
+#'   
+#'   # show results table containing additional annotations
+#'   res_dt
+#'   
+NULL
+
+#' @describeIn spliceTypeAnnotations This method calculates basic annotations 
+#'     based on overlap with the reference annotation (start, end, none, both) 
+#'     for the full fds. The overlap type is added as a new column 
+#'     \code{annotatedJunction} in \code{mcols(fds)}. 
+#' @export
+annotateIntronReferenceOverlap <- function(fds, txdb, BPPARAM=bpparam()){
+    print("loading introns")
+    #seqlevelsStyle(fds) <- seqlevelsStyle(txdb)[1]
+    introns <- unique(unlist(intronsByTranscript(txdb)))
+    # reduce the introns to only the actually expressed introns
+    fds_known <- fds[unique(to(findOverlaps(introns, 
+                                rowRanges(fds, type = "j"), type = "equal"))),]
+    anno_introns <- as.data.table(rowRanges(fds_known, 
+                                type="j"))[,.(seqnames, start, end, strand)]
+    
+    # calculate extra columns with mean/median intron expression count
+    # add the new columns
+    print("adding median count to introns")
+    sampleCounts <- K(fds_known, type = "j")
+    anno_introns[, meanCount := rowMeans(sampleCounts)]
+    anno_introns[, medianCount := rowMedians(as.matrix(sampleCounts))]
+    # order by medianCount (highest first)
+    setorderv(anno_introns, "medianCount", order=-1) 
+    anno_introns_ranges <- makeGRangesFromDataFrame(anno_introns, 
+                                                    keep.extra.columns = TRUE)
+    
+    # get all fds junctions
+    fds_junctions <- rowRanges(fds, type = "j")
+    
+    # Do the annotation just for the intron with highest median expression
+    print("start calculating annotations")    
+    overlaps <- findOverlaps(fds_junctions, anno_introns_ranges, select="first")
+    annotations <- bplapply(seq_len(length(fds_junctions)), function(i){
```

Comment:
this might fail if you are using slurm or cluster as backend. As then the objects are not passed on to the function. So its always best practice to pass on any object needed within the function

Reply:
And anyway, is it not smarter and more efficient to vectorize this? Running it 3 times for all types and then run your tests but vectorized. e.g for `none` those the index of the junction is within the overlap hit?

Reply:
I agree that vectorized would be nicer. I had started to work on such a version of this last month, but then realized that I don't get exactly the same as Karoline had (as she considers only overlap with highest expressed annotated intron), so I sticked with this for now. Let's discuss this again, as I think changes here could lead to having to adapt in the spliceType label assignment